platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -0,0 +1,545 @@
+/*
+ * Copyright (C) 2021-2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.merkledb.collections;
+
+import static com.swirlds.common.utility.Units.MEBIBYTES_TO_BYTES;
+import static com.swirlds.merkledb.utilities.MerkleDbFileUtils.readFromFileChannel;
+
+import com.swirlds.merkledb.utilities.MerkleDbFileUtils;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.stream.LongStream;
+import java.util.stream.StreamSupport;
+import sun.misc.Unsafe;
+
+/**
+ * Common parent class for long list implementations. It takes care of loading a snapshot from disk,
+ * chunk management and other common functionality.
+ *
+ * @param <C> a type that represents a chunk (byte buffer, array or long that represents an offset of the chunk)
+ */
+public abstract class AbstractLongList<C> implements LongList {
+
+    /** Access to sun.misc.Unsafe required for operations on direct bytebuffers*/
+    protected static final Unsafe UNSAFE;
+
+    static {
+        try {
+            final Field f = Unsafe.class.getDeclaredField("theUnsafe");
+            f.setAccessible(true);
+            UNSAFE = (Unsafe) f.get(null);
+        } catch (final NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
+            throw new InternalError(e);
+        }
+    }
+
+    /** A suitable default for the maximum number of longs that may be stored (32GB of longs). */
+    protected static final long DEFAULT_MAX_LONGS_TO_STORE = 4_000_000_000L;
+    /** The maximum number of longs to store per chunk. */
+    protected static final int MAX_NUM_LONGS_PER_CHUNK = Math.toIntExact(16_000L * (MEBIBYTES_TO_BYTES / Long.BYTES));
```

Comment:
Not for this PR, but it might be worth considering if this value and some of the other sizing parameters here should be read from configuration.

Reply:
Sure, I'll make it configurable in the follow up PR